src/Features/Core/Portable/SplitOrMergeIfStatements/Nested/AbstractMergeNestedIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,199 @@
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeNestedIfStatementsCodeRefactoringProvider
+        : AbstractMergeIfStatementsCodeRefactoringProvider
+    {
+        // Converts:
+        //    if (a)
+        //    {
+        //        if (b)
+        //            Console.WriteLine();
+        //    }
+        //
+        // To:
+        //    if (a && b)
+        //        Console.WriteLine();
+
+        protected sealed override CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument, string ifKeywordText)
+            => new MyCodeAction(createChangedDocument, ifKeywordText);
+
+        protected sealed override async Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifStatement, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var ifGenerator = document.GetLanguageService<IIfLikeStatementGenerator>();
+
+            return IsFirstStatementOfIfLikeStatement(syntaxFacts, ifGenerator, ifStatement, out var outerIfLikeStatement) &&
+                   await CanBeMergedWithOuterAsync(document, syntaxFacts, ifGenerator, outerIfLikeStatement, ifStatement, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected sealed override SyntaxNode GetChangedRoot(Document document, SyntaxNode root, SyntaxNode ifStatement)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var ifGenerator = document.GetLanguageService<IIfLikeStatementGenerator>();
+            var generator = document.GetLanguageService<SyntaxGenerator>();
+
+            Contract.ThrowIfFalse(IsFirstStatementOfIfLikeStatement(syntaxFacts, ifGenerator, ifStatement, out var outerIfLikeStatement));
+
+            var newCondition = generator.LogicalAndExpression(
+                ifGenerator.GetCondition(outerIfLikeStatement),
+                ifGenerator.GetCondition(ifStatement));
+
+            var newIfLikeStatement = ifGenerator.WithStatementsOf(
+                ifGenerator.WithCondition(outerIfLikeStatement, newCondition),
+                ifStatement);
+
+            return root.ReplaceNode(outerIfLikeStatement, newIfLikeStatement.WithAdditionalAnnotations(Formatter.Annotation));
+        }
+
+        private bool IsFirstStatementOfIfLikeStatement(
+            ISyntaxFactsService syntaxFacts,
+            IIfLikeStatementGenerator ifGenerator,
+            SyntaxNode statement,
+            out SyntaxNode ifLikeStatement)
+        {
+            // Check whether the statement is a first statement inside an if statement.
+            // If it's inside a block, it has to be the first statement of the block.
+
+            // This is a defensive check that should always succeed.
```

Comment:
consider asserting then if it fails.

Reply:
➡️ Fixed in 2106f9e80a49cc25622a8f14053574872b47d7dc